rslib/src/collection/backup.rs

```diff
@@ -14,34 +14,51 @@ use itertools::Itertools;
 use log::error;
 
 use crate::{
-    backend_proto::preferences::Backups, import_export::package::export_colpkg_from_data, log,
+    backend_proto::preferences::BackupLimits, import_export::package::export_colpkg_from_data, log,
     prelude::*,
 };
 
 const BACKUP_FORMAT_STRING: &str = "backup-%Y-%m-%d-%H.%M.%S.colpkg";
-/// Default seconds after a backup, in which further backups will be skipped.
-const MINIMUM_BACKUP_INTERVAL: u64 = 5 * 60;
 
-pub fn backup(
-    col_path: impl AsRef<Path>,
+pub fn maybe_backup(
+    col: &mut Collection,
     backup_folder: impl AsRef<Path> + Send + 'static,
-    limits: Backups,
-    minimum_backup_interval: Option<u64>,
-    log: Logger,
-    tr: I18n,
-) -> Result<Option<JoinHandle<()>>> {
-    let recent_secs = minimum_backup_interval.unwrap_or(MINIMUM_BACKUP_INTERVAL);
-    if recent_secs > 0 && has_recent_backup(backup_folder.as_ref(), recent_secs)? {
+    force: bool,
+) -> Result<Option<JoinHandle<Result<()>>>> {
+    if !col.changed_since_last_backup()? {
+        return Ok(None);
+    }
+    let limits = col.get_backup_limits();
+    if should_skip_backup(force, limits.minimum_interval_mins, backup_folder.as_ref())? {
         Ok(None)
     } else {
-        let col_data = std::fs::read(col_path)?;
+        let log = col.log.clone();
+        let tr = col.tr.clone();
+        col.storage.checkpoint()?;
+        let col_data = std::fs::read(&col.col_path)?;
+        col.update_last_backup_timestamp()?;
         Ok(Some(thread::spawn(move || {
             backup_inner(&col_data, &backup_folder, limits, log, &tr)
         })))
     }
 }
 
-fn has_recent_backup(backup_folder: &Path, recent_secs: u64) -> Result<bool> {
+fn should_skip_backup(
+    force: bool,
+    minimum_interval_mins: u32,
+    backup_folder: &Path,
+) -> Result<bool> {
+    if force {
+        Ok(false)
+    } else if minimum_interval_mins == 0 {
+        Ok(true)
```

Comment:
I believe it should be `Ok(false)` to match the previous behaviour.

Reply:
This is coming from the user config now, and my thinking here was the user could set it to 0 if they didn't want any automatic periodic/on close backups.

Reply:
This should be explained then, shouldn't it? Otherwise I'd assume, lower *minutes between backups* -> more backups.

Reply:
I'd been thinking we could do that by tweaking the description or adding a separate checkbox that sets it to 0 behind the scenes. I just tried adding the checkbox to the preferences screen, but having a single switch that could accidentally get disabled feels risky, as backups are so important. Could the following work instead?

- remove the special case for 0, but set the minimum in the spinner to 5 so the user doesn't waste disk/get confused by the 5 minute duration of the time
- increase the maximum for the minimum interval to 99999, so users can't fully disable automatic backups, but can have them created only once every 2 months

Reply:
Sounds good!

Reply:
On second thought, even 9999 is probably reasonable - once a week is not going to consume large amounts of space. Might start with that and see how many people want more (and they could always use the debug console if they know what they're doing)

Reply:
Also, there are the other options which effectively allow to have only one backup stored at any point in time.